### PR TITLE
feat(router): Move organization_id to request header from request body for v2

### PR DESCRIPTION
**Pull request by AnuthaDev**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

For create_merchant request organization_id is currently being passed in request body. Moved it to request header

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

1a. Correct Request

```
curl --location 'http://localhost:8080/v2/merchant_accounts' \
--header 'X-Organization-Id: org_55QZkS5H9AR8deTk9mRs' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
    "merchant_name": "Cloth Seller Person"
}'
```

1b. Response

```json
{
    "id": "clothsellerperson_iG5VNjsN9xuCg7Xx0uWh",
    "merchant_name": "Cloth Seller Person",
    "merchant_details": null,
    "publishable_key": "pk_dev_1114028c68c242dc917bde0ec8fea75b",
    "metadata": null,
    "organization_id": "org_55QZkS5H9AR8deTk9mRs",
    "recon_status": "not_requested"
}
```

2a. Non-existant Org-ID Request
```
curl --location 'http://localhost:8080/v2/merchant_accounts' \
--header 'X-Organization-Id: org_67ZZlS5H9AR8deTk8oPi' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
    "merchant_name": "Cloth Seller Person"
    
}'
```

2b. Response

```json
{
    "error": {
        "type": "invalid_request",
        "message": "organization with the given id does not exist",
        "code": "IR_37"
    }
}
```

3a. Invalid Org_id request

```
curl --location 'http://localhost:8080/v2/merchant_accounts' \
--header 'X-Organization-Id: I dont exist' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
    "merchant_name": "Cloth Seller Person"
    
}'
```

3b. Response

```json
{
    "error": {
        "type": "invalid_request",
        "message": "`X-Organization-Id` header is invalid",
        "code": "IR_06"
    }
}
```

## Cypress Screenshot:
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/openapi/src/routes/merchant_account.rs [hunk 1]
```diff
@@ -52,13 +52,12 @@ pub async fn merchant_account_create() {}
     post,
     path = "/v2/merchant_accounts",
     request_body(
-        content = MerchantAccountCreate,
+        content = MerchantAccountCreateWithoutOrgId,
```
Comment (jarnura) [inline crates/openapi/src/routes/merchant_account.rs:55]: In documentation we shouldn't say this, it is just `MerchantAccountCreate` only

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/openapi/src/routes/merchant_account.rs (hunk 1 above)
Comment (Narayanbhat166) [reply crates/openapi/src/routes/merchant_account.rs:55]: @jarnura this won't be displayed in the openapi reference

Comment (AnuthaDev) [reply crates/openapi/src/routes/merchant_account.rs]: @jarnura Resolved. Please review.

**Comment by AnuthaDev** [comment]
Added X-Organization-Id to documentation:

image

**Review by jarnura — APPROVED** [review]

File: crates/api_models/src/admin.rs [hunk 2]
```diff
@@ -189,9 +190,19 @@ pub struct MerchantAccountCreate {
...
+// The following struct is only used internally, so we can reuse the common
+// part of `create_merchant_account` without duplicating its code for v2
+#[cfg(feature = "v2")]
+#[derive(Clone, Debug, Serialize)]
+pub struct MerchantAccountCreate {
+    pub merchant_name: Secret<common_utils::new_type::MerchantName>,
+    pub merchant_details: Option<MerchantDetails>,
+    pub metadata: Option<pii::SecretSerdeValue>,
+
+    // #[schema(value_type = String, max_length = 64, min_length = 1, example = "org_q98uSGAYbjEwqs0mJwnz")]
```
Comment (Narayanbhat166) [inline crates/api_models/src/admin.rs:205]: you can remove this commented code if it is not serving any purpose

File: crates/router/src/services/authentication.rs [hunk 3]
```diff
@@ -819,6 +819,18 @@ impl<'a> HeaderMapStruct<'a> {
                 )
             })
     }
+    #[cfg(feature = "v2")]
+    pub fn get_organization_id_from_header(&self) -> RouterResult<id_type::OrganizationId> {
+        self.get_mandatory_header_value_by_key(headers::X_ORGANIZATION_ID.into())
+            .map(|val| val.to_owned())
+            .and_then(|merchant_id| {
+                id_type::OrganizationId::wrap(merchant_id).change_context(
```
Comment (Narayanbhat166) [inline crates/router/src/services/authentication.rs:827]:
```suggestion
            .and_then(|organization_id| {
                id_type::OrganizationId::wrap(organization_id).change_context(
```

File: crates/api_models/src/admin.rs [hunk 4]
```diff
@@ -178,7 +178,8 @@ impl MerchantAccountCreate {
 #[cfg(feature = "v2")]
 #[derive(Clone, Debug, Deserialize, ToSchema, Serialize)]
 #[serde(deny_unknown_fields)]
-pub struct MerchantAccountCreate {
+#[schema(as = MerchantAccountCreate)]
```
Comment (Narayanbhat166) [inline crates/api_models/src/admin.rs:181]: What does this do? The MerchantAccountCreate does not derive ToSchema

Comment (AnuthaDev) [reply crates/api_models/src/admin.rs:181]: This is required to rename `MerchantAccountCreateWithoutOrgId` to `MerchantAccountCreate` in the openapi.json file

**Review by apoorvdixit88 — DISMISSED** [review]
Dashboard changes look fine.

**Review by Narayanbhat166 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by SanchithHegde — DISMISSED** [review]

**Review by likhinbopanna — DISMISSED** [review]

**Review by apoorvdixit88 — APPROVED** [review]
Looks good for dashboard changes.

**Comment by Gnanasundari24** [comment]
@Anurag-05-prog Run the v2 cypress test for organization and add the screenshot

**Comment by AnuthaDev** [comment]
@Gnanasundari24 Added

**Review by apoorvdixit88 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by likhinbopanna — APPROVED** [review]

**Pull request merged**